### PR TITLE
Remove cs_main from ThreadMnbRequestConnections

### PR DESCRIPTION
**Pull request by UdjinM6**

### CODE REVIEW

File: src/net.cpp [hunk 1]
```diff
@@ -1858,13 +1858,14 @@ void CConnman::ThreadMnbRequestConnections()
...
-            pnode->AddRef();
-        }
+        CNode* pnode = ConnectNode(CAddress(p.first, NODE_NETWORK), NULL, true);
+
+        // To ensure that node wasn't/won't be moved to "disconnected"
+        // and/or wasn't/won't be deleted in ThreadSocketHandler
+        LOCK(cs_vNodes);
+
+        // Too late, node was either deleted or disconnected before we locked above
+        if(!pnode || pnode->fDisconnect) continue;
```
Comment (codablock) [inline src/net.cpp]: At this point, pnode would be a stale pointer if it got deleted already. So the "!pnode" check would still be true, but fDisconnect would contain stale data.

What about ignoring the return value of ConnectNode here and do a FindNode with cs_vNodes locked afterwards. This way the lock is only held in a very short timespan and when checking the return value of FindNode, you can be sure it did not get deleted. An alternative would be to use ForNode instead, but I'm not sure about where the grant.MoveTo had to be put in that case, as I didn't check what it is for.

File: src/net.cpp [hunk 2]
```diff
@@ -1858,13 +1858,14 @@ void CConnman::ThreadMnbRequestConnections()
         std::pair<CService, std::set<uint256> > p = mnodeman.PopScheduledMnbRequestConnection();
         if(p.first == CService() || p.second.empty()) continue;
 
-        CNode* pnode = NULL;
-        {
-            LOCK2(cs_main, cs_vNodes);
-            pnode = ConnectNode(CAddress(p.first, NODE_NETWORK), NULL, true);
```
Comment (codablock) [inline src/net.cpp]: Not part of this PR, but as far as I understand we leak node refs here. ConnectNode does an AddRef here no matter if it's a new connection or an already existing one. This means that the first call to ConnectNode is fine here, as it results in ref=1, basically meaning that CConman exclusively owns the node, and thus will delete it at some point in the future. If we call this a second time, the returned pnode will have ref=2, and thus the node will never be deleted.

This is independent from the AddRef/Release calls we do here on our own, as it does not change the end result of the ref counter. This means that this PR would not change/fix the ref leaking.

I think this was introduced with the network refactoring, which missed the commit https://github.com/bitcoin/bitcoin/commit/236618061a445d2cb11e722cfac5fdae5be26abb. I remember that you had a discussion somewhere with @OlegGirko about omitting this change, probably due to Dash using ConnectNode in multiple places directly.

The solution would be to backport the missing commit and fix all places where Dash uses ConnectNode to use OpenNetworkConnection+ForNode. I can do this later today or tomorrow, assuming that what I see here is correct :)

File: src/net.cpp [hunk 3]
```diff
@@ -1858,13 +1858,13 @@ void CConnman::ThreadMnbRequestConnections()
         std::pair<CService, std::set<uint256> > p = mnodeman.PopScheduledMnbRequestConnection();
         if(p.first == CService() || p.second.empty()) continue;
 
-        CNode* pnode = ConnectNode(CAddress(p.first, NODE_NETWORK), NULL, true);
+        {
+            CNode* pnode = ConnectNode(CAddress(p.first, NODE_NETWORK), NULL, true);
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/net.cpp]: Because I need some coffee 😛

File: src/net.cpp (hunk 2 above)
Comment (UdjinM6) [reply src/net.cpp]: This shouldn't be the case - `AddRef` is only called if we "connect" to existing node as a masternode now and it wasn't a masternode before, or if we connect to a brand new node as a masternode.

Comment (codablock) [reply src/net.cpp]: Ahhh...I need some coffee as well :)
Ok, postponing the refactoring then and making it part of my 0.13+0.14 backports.

**Comment by codablock** [comment]
utACK

**Pull request merged**